src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +255,166 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
+static bool
+streamBufferedXID(ReplayStreamCtx *replayCtx)
+{
+	/* Open replaybuffer in read-only mode to use it as input stream */
+	FILE *txnBufferForRead = fopen_read_only(replayCtx->currentTxnFileName);
+
+	if (txnBufferForRead == NULL)
+	{
+		log_error("Failed to open transaction buffer file \"%s\" in "
+				  "readonly mode: %m",
+				  replayCtx->currentTxnFileName);
+		return false;
+	}
+
+	ReadFromStreamContext readerContext = {
+		.callback = stream_replay_line,
+		.ctx = replayCtx
+	};
+
+	if (!read_from_stream(txnBufferForRead, &readerContext))
+	{
+		log_error("Failed to read from replay buffer, "
+				  "see above for details");
+		return false;
+	}
+
+	fclose(txnBufferForRead);
+
+	return true;
+}
+
+
+/* streamReplayLineIntoBuffer is a callback function for the ReadFromStreamContext
+ * and read_from_stream infrastructure. It's called on each line read from a
+ * stream such as a unix pipe and buffers the line into a file until we encounter
+ * a COMMIT. This ensures that we can replay the whole transaction in one go
+ * and avoids deadlock when the BEGIN messages arrives before the existence of
+ * transation metadata file which is created by the transform process.
+ */
+static bool
+streamReplayLineIntoBuffer(void *ctx, const char *line, bool *stop)
+{
+	ReplayStreamCtx *replayCtx = (ReplayStreamCtx *) ctx;
+	StreamApplyContext *context = &(replayCtx->applyContext);
+
+	LogicalMessageMetadata metadata = { 0 };
+
+	if (!parseSQLAction((char *) line, &metadata))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/* We don't use this metadata in this function */
+	free(metadata.jsonBuffer);
+
+	if (metadata.action == STREAM_ACTION_BEGIN)
+	{
+		if (replayCtx->isTranscationBuffering)
+		{
+			log_debug("Received %s when transaction is already "
+					  "in buffering mode, ignoring.", line);
+			fclose(replayCtx->txnBuffer);
+		}
+
+		sformat(replayCtx->currentTxnFileName, MAXPGPATH, "%s/%d.sql",
+				context->paths.dir, metadata.xid);
+
+		/*
+		 * Open current transaction file in w+ mode to truncate if the file
+		 * already exists.
+		 */
+		replayCtx->txnBuffer = fopen_with_umask(replayCtx->currentTxnFileName,
+												"w+",
+												O_RDWR | O_TRUNC | O_CREAT,
+												0644);
+		if (replayCtx->txnBuffer == NULL)
+		{
+			log_error("Failed to open transaction buffer file \"%s\": %m",
+					  replayCtx->currentTxnFileName);
+			return false;
+		}
+
+		/* Create as a symlink to the current transcation file */
+		if (!create_symbolic_link(replayCtx->currentTxnFileName,
+								  context->paths.txnlatestfile))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		fformat(replayCtx->txnBuffer, "%s\n", line);
+
+		replayCtx->isTranscationBuffering = true;
+	}
+	else if (metadata.action == STREAM_ACTION_COMMIT)
+	{
+		if (!replayCtx->isTranscationBuffering)
+		{
+			log_debug("Received %s when transaction is not "
+					  "in buffering mode, ignoring.", line);
+			return true;
+		}
```

Comment:
Maybe that's when we could know that a transaction is in progress even though we didn't read the BEGIN message in this session (resuming transform case etc).

Reply:
Consider there are 2 sql files, A.sql & B.sql. B.sql has a continued transaction and the `catchup` process has exited because there is no C.sql available to apply. In the meantime C.sql was written partially by the `transform`. After switching to `replay` mode, we recognise that C.sql exists and try to apply it. But this is not going to work because  B.sql was executed in `prefetch` SQL connection/session which is no longer available in `replay` mode. Does it really makes sense to resume with the partial C.sql ?

Reply:
Thanks for the comments in the source code. See my other review comments here for further discussion.